Curl script one line installer and Debian, CentOS support.

xavierchanth opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines in CONTRIBUTING.md

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

1. I created a script that can be curled with:  
```
curl -fsSL https://raw.githubusercontent.com/XavierChanth/dess/curl-testing/getdess.sh | bash
```
(The instructions.md file is updated with the correct url for atsign-foundation's trunk branch)  
Progress towards #6, still need a more formal github page to close it.

2. Commands are added to the path:  
Before:  
```
./create.sh
#OR
./reshowqr.sh
```
After:  
```
dess-create
#OR
dess-reshowqr
```

3. Added support for other operating systems
Should support Ubuntu, Debian, Centos7, Centos8, Fedora.

Tested and confirmed:
- Ubuntu 20.04
- Debian 10
- Centos 7

**- How I did it**

- Added a detection function to use the correct installation method based on the distro
- Switched to commands available on all of the distros listed above
- Used fallback functions to support distros with varying commands/flags
- Installing fuse and squashfuse on virtualized distros to make sure snapd installs correctly
- Curl all required files to remove git as a dependency
- Combine install_software.sh and scripts/atsign.sh into one file, getdess.sh
- Wrapped everything in functions to prevent bad installations if the curl command is interrupted

**- How to verify it**

- Run the curl command above to pull it from my fork
- Use dess-create command to spin up a dess instance

**- Description for the changelog**
Debian and CentOS support, and one-line installer with curl.
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

xavierchanth committed:
add getdess.sh

xavierchanth committed:
update dependency list

xavierchanth committed:
outline curl install script

xavierchanth committed:
Main install function

xavierchanth committed:
Basic helpers for the script

xavierchanth committed:
preinstall system detection function

xavierchanth committed:
Dev note

xavierchanth committed:
dependency installer

xavierchanth committed:
install certbot script

xavierchanth committed:
install docker function

xavierchanth committed:
add compose url

xavierchanth committed:
setup atsign user function

xavierchanth committed:
setup docker function

xavierchanth committed:
export vars to internal session

xavierchanth committed:
atsign user docker test function

xavierchanth committed:
add get dess scripts function

xavierchanth committed:
add sh_c notes

xavierchanth committed:
use openssl for hashing

xavierchanth committed:
fix getdess.sh

xavierchanth committed:
give the original user docker permissions

xavierchanth committed:
testing environment preservation

xavierchanth committed:
update for testing

xavierchanth committed:
run functions one by one

xavierchanth committed:
frozen in loop fix

xavierchanth committed:
fix path

xavierchanth committed:
shellcheck and formatting fix

xavierchanth committed:
url fix repo

xavierchanth committed:
add missing flags on curl commands

xavierchanth committed:
debug

xavierchanth committed:
direct pathing for atsign home

xavierchanth committed:
no preserve for atsign user

xavierchanth committed:
remove color change

xavierchanth committed:
exact path for shepherd file

xavierchanth committed:
exact pathing

xavierchanth committed:
exact path for docker test

xavierchanth committed:
start docker daemon and wait for it

xavierchanth committed:
use runuser for atsign test

xavierchanth committed:
use runuser to run as atsign

xavierchanth committed:
wrap command in string

xavierchanth committed:
fix script to match new run method

xavierchanth committed:
give atsign user ownership

xavierchanth committed:
Update create.sh

Remove bad flag

xavierchanth committed:
shellcheck fix

xavierchanth committed:
use sh_c fallback instead of sudo

xavierchanth committed:
Merge branch 'curl-testing' of github.com:XavierChanth/dess into curl-testing

xavierchanth committed:
shellcheck fix

xavierchanth committed:
add command exists function in create.sh

xavierchanth committed:
more create testing for cross-platform

xavierchanth committed:
clean up openssl output

xavierchanth committed:
add restart script to curl install

xavierchanth committed:
update create.sh

xavierchanth committed:
wrong base-url for docker-compose

xavierchanth committed:
cross platform fixes

xavierchanth committed:
fix pathing issues

xavierchanth committed:
update create.sh

xavierchanth committed:
update create.sh

xavierchanth committed:
guarantee snapd starts correctly

xavierchanth committed:
path fix

xavierchanth committed:
Update repo url

xavierchanth committed:
Update Instructions.md

curl url is currently pointing to the raw content of github.

xavierchanth committed:
Update Instructions.md

xavierchanth commented:
Apologies for the big PR, many micro-adjustments were made to get full support across all distros.

cpswan commented:
Trying this on Amazon Linux...

```
[ec2-user@ip-172-31-11-244 ~]$ curl -fsSL https://raw.githubusercontent.com/XavierChanth/dess/curl-testing/getdess.sh | bash
Detected distribution: amazon
Your distribution is not supported by this script.
```

So I:

```
[ec2-user@ip-172-31-11-244 ~]$ curl -fsSL https://raw.githubusercontent.com/XavierChanth/dess/curl-testing/getdess.sh > getdess.sh
[ec2-user@ip-172-31-11-244 ~]$ vi getdess.sh
[ec2-user@ip-172-31-11-244 ~]$ bash getdess.sh
```

vi is used to add `amazon` to `redhat_releases='centos fedora'`

But:

```
Detected distribution: amazon
Detected package manager: yum
sh: install_dependencies: command not found
sh: install_certbot: command not found
sh: install_docker: command not found
sh: setup_atsign_user: command not found
sh: setup_docker: command not found
sh: test_atsign_user: command not found
sh: get_dess_scripts: command not found
```

cpswan commented:
Similar problem with Ubuntu:

```
ubuntu@ip-172-31-10-253:~$ curl -fsSL https://raw.githubusercontent.com/XavierChanth/dess/curl-testing/getdess.sh | bash
Detected distribution: ubuntu
Detected package manager: apt-get
sh: 1: install_dependencies: not found
sh: 1: install_certbot: not found
sh: 1: install_docker: not found
sh: 1: setup_atsign_user: not found
sh: 1: setup_docker: not found
sh: 1: test_atsign_user: not found
sh: 1: get_dess_scripts: not found
```

cpswan reviewed (changes requested):
Testing on Amazon Linux and an AWS EC2 Ubuntu instance hits issues with functions not being found.

cpswan commented on getdess.sh:
```diff
@@ -0,0 +1,280 @@
+  if [ -z "$os_release" ]
+  then
+      echo 'Error: Could not detect your distribution.'
+      exit 0
```
Consider non 0 exit codes for failure

XavierChanth replied:
I put a safe guard in to test against /etc/os-release, some distros need additional packages to run snapd. I will spin up some EC2 instances and do some testing.

xavierchanth replied:
Made this change

xavierchanth commented:
Okay, so I realize that, the sudo -E flag will not preserve the environment if changing users (in the script, non-root to root).
This means that the script should work as is if we pipe to sudo bash instead:
```
curl -fsSL https://raw.githubusercontent.com/XavierChanth/dess/curl-testing/getdess.sh | sudo bash
```
The only problem is that we won't be able to add the current user to the docker usergroup in the script.
```
sudo -E bash -c 'echo $USER'
> root
```

xavierchanth committed:
Change exit to non-zero on error

xavierchanth committed:
added amazon to redhat list

xavierchanth committed:
add pipe to sudo bash in instructions

xavierchanth committed:
add non-root guard

xavierchanth committed:
revert repo_url (this needs to be changed post-pr)

cpswan commented:
I just tried again with an Amazon Linux instance. Seems it's trying to use snap, which isn't going to work on a RedHat derived OS:

```
[ec2-user@ip-172-31-4-47 ~]$ curl -fsSL https://raw.githubusercontent.com/XavierChanth/dess/curl-testing/getdess.sh | sudo bash
Detected distribution: amazon
Detected package manager: yum
Loaded plugins: extras_suggestions, langpacks, priorities, update-motd
No packages marked for update
Loaded plugins: extras_suggestions, langpacks, priorities, update-motd
Package curl-7.61.1-12.amzn2.0.2.aarch64 already installed and latest version
Nothing to do
Loaded plugins: extras_suggestions, langpacks, priorities, update-motd
Package 1:openssl-1.0.2k-19.amzn2.0.6.aarch64 already installed and latest version
Nothing to do
Loaded plugins: extras_suggestions, langpacks, priorities, update-motd
Resolving Dependencies
--> Running transaction check
---> Package qrencode.aarch64 0:3.4.1-3.amzn2.0.2 will be installed
--> Finished Dependency Resolution

Dependencies Resolved

================================================================================
 Package        Arch          Version                    Repository        Size
================================================================================
Installing:
 qrencode       aarch64       3.4.1-3.amzn2.0.2          amzn2-core        19 k

Transaction Summary
================================================================================
Install  1 Package

Total download size: 19 k
Installed size: 69 k
Downloading packages:
qrencode-3.4.1-3.amzn2.0.2.aarch64.rpm                     |  19 kB   00:00
Running transaction check
Running transaction test
Transaction test succeeded
Running transaction
  Installing : qrencode-3.4.1-3.amzn2.0.2.aarch64                           1/1
  Verifying  : qrencode-3.4.1-3.amzn2.0.2.aarch64                           1/1

Installed:
  qrencode.aarch64 0:3.4.1-3.amzn2.0.2

Complete!
Loaded plugins: extras_suggestions, langpacks, priorities, update-motd
No package snapd available.
Error: Nothing to do
Failed to execute operation: No such file or directory
Failed to start snapd.service: Unit not found.
```

xavierchanth commented:
It is installing the certbot snap, I was able to get centos and fedora to install snapd after installing a few extra packages, that may be the case here too.

cpswan commented:
I think using snapd to install stuff on RH derived builds falls into 'unnatural acts' (and I must confess I'm not **that** comfortable with it even on Ubuntu). Much better to use an approach that cleanly uses yum (maybe even ditch snap in favour of apt for the deb stuff).

xavierchanth commented:
Ok, I will try to remove snapd as a dependency, I will replace it with the apt/yum installations for certbot.

xavierchanth committed:
remove unnecessary logic

xavierchanth committed:
change to non snap path for certbot

xavierchanth committed:
remove whitespace

xavierchanth committed:
switch to case for certbot os test

xavierchanth committed:
add exit clause for systemd check

xavierchanth committed:
add version notes

xavierchanth committed:
remove lxc support packages for snapd

xavierchanth committed:
remove unused function exports

xavierchanth committed:
add amazon linux certbot support

xavierchanth committed:
change branch of repo_url

xavierchanth committed:
typo correction

xavierchanth committed:
add equals to get only ID values for os-release

xavierchanth committed:
install docker using amazon-linux-extras for amzn

xavierchanth committed:
only use major for rhel epel link

xavierchanth committed:
Merge pull request #1 from XavierChanth/certbot-install-testing

Certbot install testing

xavierchanth commented:
Hey @cpswan please give it another go, I was able to get this up and running on Amazon Linux and Ubuntu on EC2. The overall installation should be more consistent across distro versions, I changed getting /etc/os-release from NAME to use ID instead. I also took the necessary steps to install docker and certbot via the amazon-linux-extras tool.

xavierchanth committed:
Update getdess.sh

cpswan commented:
Here's the full output from my test run on an Amazon Linux VM:

```
[ec2-user@ip-172-31-7-83 ~]$ curl https://raw.githubusercontent.com/XavierChanth/dess/curl-testing/getdess.sh | sudo bash
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  6040  100  6040    0     0  30351      0 --:--:-- --:--:-- --:--:-- 30351
Detected release id: amzn, version: 2
Detected package manager: yum
Loaded plugins: extras_suggestions, langpacks, priorities, update-motd
No packages marked for update
Loaded plugins: extras_suggestions, langpacks, priorities, update-motd
Resolving Dependencies
--> Running transaction check
---> Package qrencode.aarch64 0:3.4.1-3.amzn2.0.2 will be installed
--> Finished Dependency Resolution

Dependencies Resolved

================================================================================
 Package        Arch          Version                    Repository        Size
================================================================================
Installing:
 qrencode       aarch64       3.4.1-3.amzn2.0.2          amzn2-core        19 k

Transaction Summary
================================================================================
Install  1 Package

Total download size: 19 k
Installed size: 69 k
Downloading packages:
qrencode-3.4.1-3.amzn2.0.2.aarch64.rpm                     |  19 kB   00:00
Running transaction check
Running transaction test
Transaction test succeeded
Running transaction
  Installing : qrencode-3.4.1-3.amzn2.0.2.aarch64                           1/1
  Verifying  : qrencode-3.4.1-3.amzn2.0.2.aarch64                           1/1

Installed:
  qrencode.aarch64 0:3.4.1-3.amzn2.0.2

Complete!
Loaded plugins: extras_suggestions, langpacks, priorities, update-motd
No package certbot available.
Error: Nothing to do
Installing docker
Loaded plugins: extras_suggestions, langpacks, priorities, update-motd
Existing lock /var/run/yum.pid: another copy is running as pid 1345.
Another app is currently holding the yum lock; waiting for it to exit...
  The other application is: yum
    Memory :  99 M RSS (119 MB VSZ)
    Started: Wed Jun 23 09:20:27 2021 - 00:01 ago
    State  : Running, pid: 1345
Cleaning repos: amzn2-core amzn2extra-docker
12 metadata files removed
4 sqlite files removed
0 metadata files removed
Loaded plugins: extras_suggestions, langpacks, priorities, update-motd
amzn2-core                                               | 3.7 kB     00:00
amzn2extra-docker                                        | 3.0 kB     00:00
(1/5): amzn2-core/2/aarch64/group_gz                       | 2.5 kB   00:00
(2/5): amzn2extra-docker/2/aarch64/updateinfo              |   76 B   00:00
(3/5): amzn2extra-docker/2/aarch64/primary_db              |  57 kB   00:00
(4/5): amzn2-core/2/aarch64/updateinfo                     | 373 kB   00:00
(5/5): amzn2-core/2/aarch64/primary_db                     |  36 MB   00:00
Resolving Dependencies
--> Running transaction check
---> Package docker.aarch64 0:20.10.4-1.amzn2 will be installed
--> Processing Dependency: runc >= 1.0.0 for package: docker-20.10.4-1.amzn2.aarch64
--> Processing Dependency: libcgroup >= 0.40.rc1-5.15 for package: docker-20.10.4-1.amzn2.aarch64
--> Processing Dependency: containerd >= 1.3.2 for package: docker-20.10.4-1.amzn2.aarch64
--> Processing Dependency: pigz for package: docker-20.10.4-1.amzn2.aarch64
--> Running transaction check
---> Package containerd.aarch64 0:1.4.4-1.amzn2 will be installed
---> Package libcgroup.aarch64 0:0.41-21.amzn2 will be installed
---> Package pigz.aarch64 0:2.3.4-1.amzn2.0.1 will be installed
---> Package runc.aarch64 0:1.0.0-0.3.20210225.git12644e6.amzn2 will be installed
--> Finished Dependency Resolution

Dependencies Resolved

================================================================================
 Package    Arch    Version                             Repository         Size
================================================================================
Installing:
 docker     aarch64 20.10.4-1.amzn2                     amzn2extra-docker  25 M
Installing for dependencies:
 containerd aarch64 1.4.4-1.amzn2                       amzn2extra-docker  18 M
 libcgroup  aarch64 0.41-21.amzn2                       amzn2-core         67 k
 pigz       aarch64 2.3.4-1.amzn2.0.1                   amzn2-core         77 k
 runc       aarch64 1.0.0-0.3.20210225.git12644e6.amzn2 amzn2extra-docker 2.8 M

Transaction Summary
================================================================================
Install  1 Package (+4 Dependent packages)

Total download size: 45 M
Installed size: 202 M
Downloading packages:
(1/5): libcgroup-0.41-21.amzn2.aarch64.rpm                 |  67 kB   00:00
(2/5): pigz-2.3.4-1.amzn2.0.1.aarch64.rpm                  |  77 kB   00:00
(3/5): containerd-1.4.4-1.amzn2.aarch64.rpm                |  18 MB   00:00
(4/5): runc-1.0.0-0.3.20210225.git12644e6.amzn2.aarch64.rp | 2.8 MB   00:00
(5/5): docker-20.10.4-1.amzn2.aarch64.rpm                  |  25 MB   00:01
--------------------------------------------------------------------------------
Total                                               35 MB/s |  45 MB  00:01
Running transaction check
Running transaction test
Transaction test succeeded
Running transaction
  Installing : runc-1.0.0-0.3.20210225.git12644e6.amzn2.aarch64             1/5
  Installing : containerd-1.4.4-1.amzn2.aarch64                             2/5
  Installing : pigz-2.3.4-1.amzn2.0.1.aarch64                               3/5
  Installing : libcgroup-0.41-21.amzn2.aarch64                              4/5
  Installing : docker-20.10.4-1.amzn2.aarch64                               5/5
  Verifying  : libcgroup-0.41-21.amzn2.aarch64                              1/5
  Verifying  : pigz-2.3.4-1.amzn2.0.1.aarch64                               2/5
  Verifying  : docker-20.10.4-1.amzn2.aarch64                               3/5
  Verifying  : containerd-1.4.4-1.amzn2.aarch64                             4/5
  Verifying  : runc-1.0.0-0.3.20210225.git12644e6.amzn2.aarch64             5/5

Installed:
  docker.aarch64 0:20.10.4-1.amzn2

Dependency Installed:
  containerd.aarch64 0:1.4.4-1.amzn2
  libcgroup.aarch64 0:0.41-21.amzn2
  pigz.aarch64 0:2.3.4-1.amzn2.0.1
  runc.aarch64 0:1.0.0-0.3.20210225.git12644e6.amzn2

Complete!
  0  ansible2                 available    [ =2.4.6  =2.8  =stable ]
  1  httpd_modules            available    [ =1.0  =stable ]
  2  memcached1.5             available    \
        [ =1.5.1  =1.5.16  =1.5.17 ]
  4  postgresql9.6            available    [ =9.6.8  =stable ]
  5  postgresql10             available    [ =10  =stable ]
  7  R3.4                     available    [ =3.4.3  =stable ]
  8  rust1                    available    [ =stable ]
 11  php7.2                   available    \
        [ =7.2.13  =7.2.14  =7.2.16  =7.2.17  =7.2.19  =7.2.21
          =7.2.22  =7.2.23  =7.2.24  =7.2.26  =stable ]
 13  lamp-mariadb10.2-php7.2  available    \
        [ =10.2.10_7.2.11  =10.2.10_7.2.14  =10.2.10_7.2.16
          =10.2.10_7.2.17  =10.2.10_7.2.19  =10.2.10_7.2.22
          =10.2.10_7.2.23  =10.2.10_7.2.24  =stable ]
 14  libreoffice              available    [ =5.3.6.1  =stable ]
 16  docker=latest            enabled      \
        [ =18.06.1  =18.09.9  =stable ]
 17  mate-desktop1.x          available    [ =stable ]
 18  GraphicsMagick1.3        available    \
        [ =1.3.29  =1.3.32  =1.3.34  =stable ]
 19  tomcat8.5                available    \
        [ =8.5.32  =8.5.38  =8.5.40  =8.5.42  =8.5.50  =stable ]
 20  epel                     available    [ =7.11  =stable ]
 21  testing                  available    [ =1.0  =stable ]
 22  ecs                      available    [ =stable ]
 23  corretto8                available    \
        [ =1.8.0_202  =1.8.0_212  =1.8.0_222  =1.8.0_232  =1.8.0_242
          =stable ]
 24  golang1.11               available    \
        [ =1.11.3  =1.11.11  =1.11.13  =stable ]
 25  squid4                   available    [ =4  =stable ]
 26  php7.3                   available    \
        [ =7.3.2  =7.3.3  =7.3.4  =7.3.6  =7.3.8  =7.3.9  =7.3.10
          =7.3.11  =7.3.13  =stable ]
 27  java-openjdk11           available    [ =11  =stable ]
 28  lynis                    available    [ =stable ]
 29  kernel-ng                available    [ =stable ]
 30  BCC                      available    [ =0.x  =stable ]
 31  nginx1                   available    [ =stable ]
 32  ruby2.6                  available    [ =2.6  =stable ]
 33  mock                     available    [ =stable ]
 34  postgresql11             available    [ =11  =stable ]
 35  php7.4                   available    [ =stable ]
 36  python3.8                available    [ =stable ]
 37  lustre2.10               available    [ =stable ]
 38  haproxy2                 available    [ =stable ]
 39  collectd                 available    [ =stable ]
 40  R4                       available    [ =stable ]
 41  kernel-5.4               available    [ =stable ]
 42  selinux-ng               available    [ =stable ]
 43  php8.0                   available    [ =stable ]
 44  tomcat9                  available    [ =stable ]
 45  unbound1.13              available    [ =stable ]
 46  mariadb10.5              available    [ =stable ]
 47  kernel-5.10              available    [ =stable ]
 48  redis6                   available    [ =stable ]
curl: (22) The requested URL returned error: 404
chmod: cannot access ‘/usr/local/bin/docker-compose’: No such file or directory
Created symlink from /etc/systemd/system/multi-user.target.wants/docker.service to /usr/lib/systemd/system/docker.service.
setting up certbot
rm: cannot remove ‘/etc/letsencrypt/*’: No such file or directory
rmdir: failed to remove ‘/etc/letsencrypt/’: No such file or directory
Creating some base directories for atsign
making /home/atsign/dess
making /home/atsign/base
making /home/atsign/atsign/var
making /home/atsign/atsign/etc
making /home/atsign/atsign/logs
curling base/.env
curling base/docker-swarm.yaml
curling base/setup.sh
curling base/shepherd.yaml
curling base/restart.sh
Swarm initialized: current node (inoezesyvz8erh2x944gv8ldc) is now a manager.

To add a worker to this swarm, run the following command:

    docker swarm join --token SWMTKN-1-1g6tlknjsaejcgqm6dghy78xk2246jx9e9jqrepyr6reviirki-039r2bwbpg6vgcqflu6a26bda 172.31.7.83:2377

To add a manager to this swarm, run 'docker swarm join-token manager' and follow the instructions.

1s4bjec4w947318k5desl8jcu
Creating network secondaries_default
Creating service secondaries_shepherd
Unable to find image 'hello-world:latest' locally
latest: Pulling from library/hello-world
256ab8fe8778: Pull complete
Digest: sha256:9f6ad537c5132bcce57f7a0a20e317228d382c3cd61edae14650eec68b2b345c
Status: Downloaded newer image for hello-world:latest

Hello from Docker!
This message shows that your installation appears to be working correctly.

To generate this message, Docker took the following steps:
 1. The Docker client contacted the Docker daemon.
 2. The Docker daemon pulled the "hello-world" image from the Docker Hub.
    (arm64v8)
 3. The Docker daemon created a new container from that image which runs the
    executable that produces the output you are currently reading.
 4. The Docker daemon streamed that output to the Docker client, which sent it
    to your terminal.

To try something more ambitious, you can run an Ubuntu container with:
 $ docker run -it ubuntu bash

Share images, automate workflows, and more with a free Docker ID:
 https://hub.docker.com/

For more examples and ideas, visit:
 https://docs.docker.com/get-started/

Docker setup correctly for atsign user
```

Some slightly worrying things fly past, which might be fine, but may alarm somebody new to this:

```
No package certbot available.
Error: Nothing to do
```

```
curl: (22) The requested URL returned error: 404
chmod: cannot access ‘/usr/local/bin/docker-compose’: No such file or directory
```

When it gets to the end, the final line is:

```
Docker setup correctly for atsign user
```

No mention of `dess-create`. When I run dess-create I see:

```
Usage create.sh <@sign> <fqdn> <port> <email> <service>
Example create.sh @bob bob.example.com 6464 bob@example.com bob
```

So `create.sh` needs to be updated to `dess-create`

If I then try to run `dess-create` things go a little awry

```
[ec2-user@ip-172-31-7-83 ~]$ dess-create @sudokusufficient getdessamazon.javagone.org 6464 chris@atsign.com ss
Creating @sudokusufficient with getdessamazon.javagone.org:6464 with email chris@atsign.com with docker service name of ss
ATSIGN=@sudokusufficient
DOMAIN=getdessamazon.javagone.org
PORT=6464
EMAIL=chris@atsign.com
SECRET=2fa7e031ccab5431219728b905ced3e4549bb989077adeb5b551dc09cd37f31b7d41d25f27d1409eead24868c6992ff2f92f0c757c72ef91bf5132fca67d04a7
Getting certificates
sh: /usr/bin/certbot: No such file or directory
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0Warning: Failed to create the file
Warning: /home/atsign/atsign/etc/live/getdessamazon.javagone.org/cacert.pem:
Warning: No such file or directory
  0  196k    0   883    0     0   3294      0  0:01:01 --:--:--  0:01:01  3282
curl: (23) Failed writing body (0 != 883)
chown: cannot access ‘/home/atsign/atsign/etc/live/getdessamazon.javagone.org’: No such file or directory
chown: cannot access ‘/home/atsign/atsign/etc/archive/getdessamazon.javagone.org’: No such file or directory
cp: cannot create regular file ‘/home/atsign/atsign/etc/renewal-hooks/deploy’: No such file or directory
Starting secondary for @sudokusufficient at getdessamazon.javagone.org on port 6464 as  on Docker
sh: /usr/bin/docker-compose: No such file or directory
Top-level object must be a mapping
Your QR-Code for @sudokusufficient
```

If I then `sudo su atsign` and have a go at running `dess-create`:

```
[atsign@ip-172-31-7-83 ec2-user]$ dess-create @sudokusufficient getdessamazon.javagone.org 6464 chris@atsign.com ss
Creating @sudokusufficient with getdessamazon.javagone.org:6464 with email chris@atsign.com with docker service name of ss
atsign is not in the sudoers file.  This incident will be reported.
atsign is not in the sudoers file.  This incident will be reported.
atsign is not in the sudoers file.  This incident will be reported.
atsign is not in the sudoers file.  This incident will be reported.
atsign is not in the sudoers file.  This incident will be reported.
atsign is not in the sudoers file.  This incident will be reported.
atsign is not in the sudoers file.  This incident will be reported.
atsign is not in the sudoers file.  This incident will be reported.
atsign is not in the sudoers file.  This incident will be reported.
atsign is not in the sudoers file.  This incident will be reported.
Getting certificates

We trust you have received the usual lecture from the local System
Administrator. It usually boils down to these three things:

    #1) Respect the privacy of others.
    #2) Think before you type.
    #3) With great power comes great responsibility.

[sudo] password for atsign:
```

xavierchanth commented:
A few things:

1. docker compose is curling the wrong url when installing on arm64 
2. I missed a change in one of the commits that prevented certbot from being installed on amazon
3. dess-create runs some commands as root, so it will require the account to have root access.

xavierchanth committed:
Fix certbot installation on amazon linux 2

xavierchanth committed:
Testing aarch64

xavierchanth committed:
Update getdess.sh

xavierchanth committed:
Add Error Codes

xavierchanth committed:
Update getdess.sh

xavierchanth committed:
Update getdess.sh

xavierchanth committed:
Update getdess.sh

xavierchanth committed:
Update getdess.sh

xavierchanth committed:
Update getdess.sh

xavierchanth committed:
Switch to  docker-compose 1.29.1

This has run.sh available for aarch64

xavierchanth committed:
Update getdess.sh

xavierchanth committed:
Update getdess.sh

xavierchanth committed:
Update getdess.sh

xavierchanth committed:
Update getdess.sh

xavierchanth committed:
Update getdess.sh

xavierchanth committed:
Update getdess.sh

xavierchanth committed:
Update getdess.sh

xavierchanth committed:
Update getdess.sh

xavierchanth committed:
Update getdess.sh

xavierchanth commented:
As of now, AMD64 has good support, only Amazon Linux is supported on ARM64.  pip was required to install docker-compose reliably on Amazon Linux ARM64, I will add the steps to install it correctly other distros.

xavierchanth committed:
Update getdess.sh

xavierchanth committed:
Update getdess.sh

xavierchanth commented:
Ubuntu on Arm looks good, checking debian, fedora, centos tomorrow.

xavierchanth committed:
Update create.sh

xavierchanth committed:
Update create.sh

xavierchanth committed:
Update reshowqr.sh

xavierchanth committed:
Update getdess.sh

xavierchanth committed:
Update getdess.sh

xavierchanth committed:
Update getdess.sh

xavierchanth committed:
add rocky installation for docker

xavierchanth committed:
Update getdess.sh

xavierchanth committed:
Update getdess.sh

xavierchanth committed:
Update getdess.sh

xavierchanth committed:
Update create.sh

xavierchanth committed:
Add post_install function

xavierchanth commented:
This is what I have so far as a fix for CIS Hardened Ubuntu #7 

#### Networking
sudo iptables -A INPUT -p tcp -m tcp --dport 80 -m state --state NEW,ESTABLISHED -j ACCEPT
sudo iptables -A INPUT -p tcp -m tcp --dport 6464 -m state --state NEW,ESTABLISHED -j ACCEPT

#### Permissons
sudo chown atsign:atsign /home/atsign -R
sudo chown atsign:docker /usr/local/bin/dess-*
sudo chmod 664 /usr/local/bin/dess-*

#### Docker-compose
sudo chown root:docker /usr/local/bin/docker-compose
sudo mount /tmp -o remount,exec

cpswan commented:
I just took this for a spin on an AWS Arm instance running Amazon Linux:

```
Building wheels for collected packages: pynacl
  Building wheel for pynacl (PEP 517) ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/bin/python3 /usr/lib/python3.7/site-packages/pip/_vendor/pep517/_in_process.py build_wheel /tmp/tmpg9k5f2i3
       cwd: /tmp/pip-install-0km25kfy/pynacl
  Complete output (98 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.linux-aarch64-3.7
  creating build/lib.linux-aarch64-3.7/nacl
  copying src/nacl/__init__.py -> build/lib.linux-aarch64-3.7/nacl
  copying src/nacl/encoding.py -> build/lib.linux-aarch64-3.7/nacl
  copying src/nacl/exceptions.py -> build/lib.linux-aarch64-3.7/nacl
  copying src/nacl/hash.py -> build/lib.linux-aarch64-3.7/nacl
  copying src/nacl/hashlib.py -> build/lib.linux-aarch64-3.7/nacl
  copying src/nacl/public.py -> build/lib.linux-aarch64-3.7/nacl
  copying src/nacl/secret.py -> build/lib.linux-aarch64-3.7/nacl
  copying src/nacl/signing.py -> build/lib.linux-aarch64-3.7/nacl
  copying src/nacl/utils.py -> build/lib.linux-aarch64-3.7/nacl
  creating build/lib.linux-aarch64-3.7/nacl/pwhash
  copying src/nacl/pwhash/__init__.py -> build/lib.linux-aarch64-3.7/nacl/pwhash
  copying src/nacl/pwhash/_argon2.py -> build/lib.linux-aarch64-3.7/nacl/pwhash
  copying src/nacl/pwhash/argon2i.py -> build/lib.linux-aarch64-3.7/nacl/pwhash
  copying src/nacl/pwhash/argon2id.py -> build/lib.linux-aarch64-3.7/nacl/pwhash
  copying src/nacl/pwhash/scrypt.py -> build/lib.linux-aarch64-3.7/nacl/pwhash
  creating build/lib.linux-aarch64-3.7/nacl/bindings
  copying src/nacl/bindings/__init__.py -> build/lib.linux-aarch64-3.7/nacl/bindings
  copying src/nacl/bindings/crypto_aead.py -> build/lib.linux-aarch64-3.7/nacl/bindings
  copying src/nacl/bindings/crypto_box.py -> build/lib.linux-aarch64-3.7/nacl/bindings
  copying src/nacl/bindings/crypto_core.py -> build/lib.linux-aarch64-3.7/nacl/bindings
  copying src/nacl/bindings/crypto_generichash.py -> build/lib.linux-aarch64-3.7/nacl/bindings
  copying src/nacl/bindings/crypto_hash.py -> build/lib.linux-aarch64-3.7/nacl/bindings
  copying src/nacl/bindings/crypto_kx.py -> build/lib.linux-aarch64-3.7/nacl/bindings
  copying src/nacl/bindings/crypto_pwhash.py -> build/lib.linux-aarch64-3.7/nacl/bindings
  copying src/nacl/bindings/crypto_scalarmult.py -> build/lib.linux-aarch64-3.7/nacl/bindings
  copying src/nacl/bindings/crypto_secretbox.py -> build/lib.linux-aarch64-3.7/nacl/bindings
  copying src/nacl/bindings/crypto_secretstream.py -> build/lib.linux-aarch64-3.7/nacl/bindings
  copying src/nacl/bindings/crypto_shorthash.py -> build/lib.linux-aarch64-3.7/nacl/bindings
  copying src/nacl/bindings/crypto_sign.py -> build/lib.linux-aarch64-3.7/nacl/bindings
  copying src/nacl/bindings/randombytes.py -> build/lib.linux-aarch64-3.7/nacl/bindings
  copying src/nacl/bindings/sodium_core.py -> build/lib.linux-aarch64-3.7/nacl/bindings
  copying src/nacl/bindings/utils.py -> build/lib.linux-aarch64-3.7/nacl/bindings
  running build_clib
  checking build system type... aarch64-unknown-linux-gnu
  checking host system type... aarch64-unknown-linux-gnu
  checking for a BSD-compatible install... /bin/install -c
  checking whether build environment is sane... yes
  checking for a thread-safe mkdir -p... /bin/mkdir -p
  checking for gawk... gawk
  checking whether make sets $(MAKE)... yes
  checking whether make supports nested variables... yes
  checking whether UID '0' is supported by ustar format... yes
  checking whether GID '0' is supported by ustar format... yes
  checking how to create a ustar tar archive... gnutar
  checking whether make supports nested variables... (cached) yes
  checking whether to enable maintainer-specific portions of Makefiles... no
  checking whether make supports the include directive... yes (GNU style)
  checking for gcc... gcc -pthread
  checking whether the C compiler works... no
  configure: error: in `/tmp/pip-install-0km25kfy/pynacl/build/temp.linux-aarch64-3.7':
  configure: error: C compiler cannot create executables
  See `config.log' for more details
  Traceback (most recent call last):
    File "/usr/lib/python3.7/site-packages/pip/_vendor/pep517/_in_process.py", line 280, in <module>
      main()
    File "/usr/lib/python3.7/site-packages/pip/_vendor/pep517/_in_process.py", line 263, in main
      json_out['return_val'] = hook(**hook_input['kwargs'])
    File "/usr/lib/python3.7/site-packages/pip/_vendor/pep517/_in_process.py", line 205, in build_wheel
      metadata_directory)
    File "/tmp/pip-build-env-64ej14cx/overlay/lib/python3.7/site-packages/setuptools/build_meta.py", line 222, in build_wheel
      wheel_directory, config_settings)
    File "/tmp/pip-build-env-64ej14cx/overlay/lib/python3.7/site-packages/setuptools/build_meta.py", line 207, in _build_with_temp_dir
      self.run_setup()
    File "/tmp/pip-build-env-64ej14cx/overlay/lib/python3.7/site-packages/setuptools/build_meta.py", line 150, in run_setup
      exec(compile(code, __file__, 'exec'), locals())
    File "setup.py", line 264, in <module>
      "Programming Language :: Python :: 3.8",
    File "/tmp/pip-build-env-64ej14cx/overlay/lib/python3.7/site-packages/setuptools/__init__.py", line 153, in setup
      return distutils.core.setup(**attrs)
    File "/usr/lib64/python3.7/distutils/core.py", line 148, in setup
      dist.run_commands()
    File "/usr/lib64/python3.7/distutils/dist.py", line 966, in run_commands
      self.run_command(cmd)
    File "/usr/lib64/python3.7/distutils/dist.py", line 985, in run_command
      cmd_obj.run()
    File "/tmp/pip-build-env-64ej14cx/overlay/lib/python3.7/site-packages/wheel/bdist_wheel.py", line 299, in run
      self.run_command('build')
    File "/usr/lib64/python3.7/distutils/cmd.py", line 313, in run_command
      self.distribution.run_command(command)
    File "/usr/lib64/python3.7/distutils/dist.py", line 985, in run_command
      cmd_obj.run()
    File "/usr/lib64/python3.7/distutils/command/build.py", line 135, in run
      self.run_command(cmd_name)
    File "/usr/lib64/python3.7/distutils/cmd.py", line 313, in run_command
      self.distribution.run_command(command)
    File "/usr/lib64/python3.7/distutils/dist.py", line 985, in run_command
      cmd_obj.run()
    File "setup.py", line 179, in run
      cwd=build_temp,
    File "/usr/lib64/python3.7/subprocess.py", line 363, in check_call
      raise CalledProcessError(retcode, cmd)
  subprocess.CalledProcessError: Command '['/tmp/pip-install-0km25kfy/pynacl/src/libsodium/configure', '--disable-shared', '--enable-static', '--disable-debug', '--disable-dependency-tracking', '--with-pic', '--prefix', '/tmp/pip-install-0km25kfy/pynacl/build/temp.linux-aarch64-3.7']' returned non-zero exit status 77.
  ----------------------------------------
  ERROR: Failed building wheel for pynacl
Failed to build pynacl
ERROR: Could not build wheels for pynacl which use PEP 517 and cannot be installed directly
1
Error: unable to install docker compose
```

xavierchanth committed:
Update getdess.sh

xavierchanth commented:
Oh boy, Amazon on Arm is quickly becoming my new best friend. I'll do some more investigation.

xavierchanth committed:
Add missing then statement

xavierchanth committed:
Update getdess.sh

xavierchanth commented:
gcc was missing as a dependency to build docker-compose using pip3. I've added it in the last commit. I ran the installer on two separate instances to ensure that it is running on Amazon Arm64. I've encountered no further issues.

xavierchanth committed:
Update getdess.sh

xavierchanth committed:
Update getdess.sh

64jantuma commented:
So I was testing the script of Hardened Ubuntu. I have experienced issues with creation of docker yaml file. Doing some digging around I found out that the problem is with hardened os /tmp is mounted as noexec. This can be fixed by adding temp folder for execution user in this case atsing by creating new temp folder
```mkdir $HOME/tmp```
and exporting it as TMPDIR 
```export TMPDIR=$HOME/tmp ```

The problem is documented in:
https://stackoverflow.com/questions/57796839/docker-compose-error-while-loading-shared-libraries-libz-so-1-failed-to-map-s

I tested it and it works like a charm.

xavierchanth committed:
At ~atsign/tmp

/tmp is not executable on hardened ubuntu, this change creates a tmp folder that we will point to in dess-create.sh

xavierchanth committed:
Point $TMPDIR to an executable folder

Hardened fix

xavierchanth committed:
Update Instructions.md

xavierchanth committed:
Update Instructions.md

xavierchanth committed:
Update create.sh

cpswan reviewed (approved):
LGTM
